assets/src/edit-story/components/form/numeric.js

```diff
@@ -97,48 +114,177 @@ function Numeric({
         // Increment value
         newValue =
           typeof max !== 'undefined'
-            ? Math.min(max, Big(value).plus(diff))
-            : Big(value).plus(diff);
+            ? Math.min(max, Big(newValue).plus(diff))
+            : Big(newValue).plus(diff);
       } else if (key === 'ArrowDown') {
         // Decrement value
         newValue =
           typeof min !== 'undefined'
-            ? Math.max(min, Big(value).minus(diff))
-            : Big(value).minus(diff);
+            ? Math.max(min, Big(newValue).minus(diff))
+            : Big(newValue).minus(diff);
       }
-      onChange(parseFloat(newValue.toString()));
+
+      validateAndSubmitInput(Number(newValue), {
+        ignoreValidation: true,
+        selectContentOnUpdate: true,
+      });
     },
-    [isMultiple, float, onChange, max, value, min]
+    [
+      isMultiple,
+      float,
+      max,
+      inputValue,
+      value,
+      min,
+      canBeNegative,
+      canBeEmpty,
+      validateAndSubmitInput,
+    ]
   );
 
+  const handleEsc = useCallback(() => {
+    // Revert input value and exit input focus without
+    // triggering blur validation
+    setInputValue(value);
+    skipValidationRef.current = true;
+    inputRef.current.blur();
+  }, [value]);
+
+  const handleEnter = useCallback(() => {
+    validateAndSubmitInput(inputValue, { selectContentOnUpdate: true });
+  }, [validateAndSubmitInput, inputValue]);
+
+  const submitForm = useCallback(() => {
+    if (targetFormRef.current) {
+      targetFormRef.current.dispatchEvent(
+        new window.Event('submit', { cancelable: true })
+      );
+
+      // Resetting to null after dispatch, it'll get
+      // re-assigned on component blur
+      targetFormRef.current = null;
+    }
+  }, []);
+
   useKeyDownEffect(
-    ref,
-    { key: ['up', 'alt+up', 'down', 'alt+down'], editable: true },
+    inputRef,
+    {
+      key: ['up', 'alt+up', 'down', 'alt+down'],
+      editable: true,
+    },
     handleUpDown,
     [handleUpDown]
   );
 
-  const { focused, handleFocus, handleBlur } = useFocusAndSelect(ref);
+  useKeyDownEffect(
+    inputRef,
+    {
+      key: ['escape'],
+      editable: true,
+    },
+    handleEsc,
+    [handleEsc]
+  );
+
+  useKeyDownEffect(
+    inputRef,
+    {
+      key: ['enter'],
+      editable: true,
+    },
+    handleEnter,
+    [handleEnter]
+  );
+
+  useEffect(() => {
+    if (targetFormRef.current) {
+      submitForm();
+    }
+
+    setInputValue(value);
+  }, [submitForm, value]);
+
+  useEffect(() => {
+    if (selectInputContents.current) {
+      inputRef.current.select();
+
+      // When we want to select the content of the input
+      // we hold open the door for a slight moment to allow
+      // all the data to flush down the pipeline.
+      window.clearTimeout(selectContentsTimeout.current);
+      selectContentsTimeout.current = window.setTimeout(() => {
+        selectInputContents.current = false;
+      }, 10);
```

Comment:
This is the work-around I needed to implement so when you hit `Enter` and push up a new change, and then `value` fluctuates a bit (which causes a deselection), this timeout ensures that after the data flush, the final input is selected.  And now I don't need any changes to `designPanel.js`, so the history tracking should work like normal.